packages/schema/parse/src/version_compare/comparators.ts

```diff
@@ -0,0 +1,489 @@
+  MethodDefinition,
+  ImportedQueryDefinition,
+  ImportedEnumDefinition,
+  ImportedObjectDefinition,
+  ImportedDefinition,
+} from "../typeInfo/definitions";
+import { TypeInfo } from "../typeInfo";
+import { compareSets, SetComparisionType } from "./setUtils";
+
+export type ImportDefinition = { type: string };
+
+export enum VersionRelease {
+  PATCH,
+  MINOR,
+  MAJOR,
+}
+
+export function compare(t1: TypeInfo, t2: TypeInfo): VersionRelease {
+  let result: VersionRelease = VersionRelease.PATCH;
+  result = Math.max(result, compareEnumTypes(t1.enumTypes, t2.enumTypes));
+  result = Math.max(result, compareObjectTypes(t1.objectTypes, t2.objectTypes));
+  result = Math.max(result, compareQueryTypes(t1.queryTypes, t2.queryTypes));
+  result = Math.max(
+    result,
+    compareImportedEnumTypes(t1.importedEnumTypes, t2.importedEnumTypes)
+  );
+  result = Math.max(
+    result,
+    compareImportedObjectTypes(t1.importedObjectTypes, t2.importedObjectTypes)
+  );
+  result = Math.max(
+    result,
+    compareImportedQueryTypes(t1.importedQueryTypes, t2.importedQueryTypes)
+  );
+  return result;
+}
+
+export function compareGenericType(
+  obj1: GenericDefinition,
+  obj2: GenericDefinition
+): VersionRelease {
+  if (
+    obj1.kind === obj2.kind &&
+    obj1.name === obj2.name &&
+    obj1.type === obj2.type
+  ) {
+    if (!obj1.required && obj2.required) {
+      return VersionRelease.MAJOR;
+    }
+    return VersionRelease.PATCH;
+  }
+  return VersionRelease.MAJOR;
+}
+
+export function compareEnumType(
+  obj1: EnumDefinition,
+  obj2: EnumDefinition
+): VersionRelease {
+  if (
+    obj1.kind === obj2.kind &&
+    obj1.name === obj2.name &&
+    obj1.type === obj2.type
+  ) {
+    if (!obj1.required && obj2.required) {
+      return VersionRelease.MAJOR;
+    }
+
+    const c1: Set<string> = new Set<string>(obj1.constants);
+    const c2: Set<string> = new Set<string>(obj2.constants);
+    const result: SetComparisionType = compareSets(c1, c2);
+
+    switch (result) {
+      case SetComparisionType.EQUAL:
+        return VersionRelease.PATCH;
+      case SetComparisionType.SUBSET:
+        return VersionRelease.MINOR;
+      default:
+        return VersionRelease.MAJOR;
+    }
+  }
+  return VersionRelease.MAJOR;
+}
+
+export function compareQueryType(
+  obj1: QueryDefinition,
+  obj2: QueryDefinition
+): VersionRelease {
+  let result: VersionRelease = VersionRelease.PATCH;
+  if (
+    obj1.kind === obj2.kind &&
+    obj1.name === obj2.name &&
+    obj1.type === obj2.type
+  ) {
+    if (!obj1.required && obj2.required) {
+      return VersionRelease.MAJOR;
+    }
+    result = Math.max(result, compareMethodTypes(obj1.methods, obj2.methods));
+    result = Math.max(result, compareImportTypes(obj1.imports, obj2.imports));
+    result = Math.max(
+      result,
+      compareGenericTypes(obj1.interfaces, obj2.interfaces)
+    );
+    return result;
+  }
+  return VersionRelease.MAJOR;
+}
+
+export function compareObjectType(
```

Comment:
I like the compare function per type approach
A potential problem I see is introducing a new property on the type, but then forgetting to add it to one of these functions

Reply:
Agree but we wouldn't know how to categories and compare this newly added property and generic approach wouldn't work.

Reply:
Yea, a generic approach wouldn't.
A standard pattern is to have the compare function be in each of the types, so you would call it like this:
QueryDefinition.compare()
But, I'm not a 100% sure that it should be the responsibility of that type.
It would help in solving the problem, since it'd be a lot easier for the dev to know to update it.

Thoughts, @dOrgJelli ?

Reply:
I agree that creating compare method for every type would solve this issue. I believe from the OOP perspective it would make sense to have this as an object method. We can also leverage that inheritance property to reduce repetition in our compare functions.